---
title: "Lab4_Transcriptomics"
author: "Mendel Kwan"
date: "2025-09-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r Required_Libraries, include = FALSE}

lapply(c("docopt", "DT", "pheatmap","GenomicFeatures", "DESeq2", "edgeR", "systemPipeR", "systemPipeRdata", "BiocStyle", "GO.db", "dplyr", "tidyr", "stringr", "Rqc", "QuasR", "DT", "ape", "clusterProfiler","biomaRt","EnhancedVolcano"), require, character.only = TRUE)
library("DT")
library("Rsubread")
```
# Question 1

```{r Locate_Samples}
fq_path <- systemPipeRdata::pathList()$fastqdir
meta_data <- read.table("/home/vboxuser/R/x86_64-pc-linux-gnu-library/4.5/systemPipeRdata/extdata/param/targetsPE.txt", header = TRUE)
meta_data$FileName1 <- gsub("./data/", fq_path, meta_data$FileName1)
meta_data$FileName2<- gsub("./data/", fq_path, meta_data$FileName2)
datatable(meta_data)
```

# Question 2

```{r Quality_Control_Report}
qcReport <- rqc(fq_path, pattern = ".fastq.gz", openBrowser = FALSE)
rqcCycleQualityBoxPlot(qcReport[1:12])
rqcCycleQualityBoxPlot(qcReport[13:24])
rqcCycleQualityBoxPlot(qcReport[25:36])
rqcCycleBaseCallsLinePlot(qcReport[1:12])
rqcCycleBaseCallsLinePlot(qcReport[13:24])
rqcCycleBaseCallsLinePlot(qcReport[25:36])
```

None of the samples have much longer reads than the others.  They are all between 75-100 bases.  The shortest number of cycles is 75.

None of the samples have a significant number of N base calls.

# Question 3

```{r Quality_Control}
fastqFiles = c()
outputFiles = c()
output_path = "~/Labs/BIOL4315_Lab4/data/processed_reads/"
for(i in 1:nrow(meta_data)) {
  row <- meta_data[i, ]
  fastqFiles <- c(fastqFiles, row$FileName1, row$FileName2)
  outputFileName1 <- gsub("/home/vboxuser/R/x86_64-pc-linux-gnu-library/4.5/systemPipeRdata/extdata/fastq/", output_path, row$FileName1)
  outputFileName1 <- gsub("fastq.gz", "processed.fastq", outputFileName1)
  meta_data[i, "FileName1"] = outputFileName1
  outputFileName2 <- gsub("/home/vboxuser/R/x86_64-pc-linux-gnu-library/4.5/systemPipeRdata/extdata/fastq/", output_path, row$FileName2)
  outputFileName2 <- gsub("fastq.gz", "processed.fastq", outputFileName2)
  meta_data[i, "FileName2"] <- outputFileName2
  outputFiles <- c(outputFiles, outputFileName1, outputFileName2 )
}

preprocessReads(fastqFiles, outputFiles, nBases = 1, truncateEndBases = 3, Lpattern = "GCCCGGGTAA")
datatable(meta_data)

```

```{r Index_Reference, include = FALSE}
reference_genome <- "~/Labs/BIOL4315_Lab4/data/GCF_000001735.4_TAIR10.1_genomic.fna"
tryCatch({
  system2(command = "hisat2-build", args = c("-p","8", reference_genome, "~/Labs/BIOL4315_Lab4/data/hisat2_index/tair10.1_index"), stdout = TRUE, stderr = TRUE)
  }, error = function(e) {
paste("hisat2-build...", "Indexing failed with error:", e$message)
})
```

hisat2-build -p 8 ~/Labs/BIOL4315_Lab4/data/GCF_000001735.4_TAIR10.1_genomic.fna ~/Labs/BIOL4315_Lab4/data/hisat2_index/tair10.1_index

# Question 4
```{r Alignment}
for(i in 1:nrow(meta_data)) {
row <- meta_data[i, ]
tryCatch({
  hisat2_output <- system2(command = "hisat2", args = c("-x","~/Labs/BIOL4315_Lab4/data/hisat2_index/tair10.1_index", row$FileName1, "-p", "8", "-S", paste("~/Labs/BIOL4315_Lab4/output/sam_files/", row$SampleName, ".sam", sep = "" )), stdout = TRUE, stderr = TRUE)
  writeLines(hisat2_output, paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, "_hisat2.log", sep = "") )
  system2(command = "samtools", args = c("view", "-bS", paste("~/Labs/BIOL4315_Lab4/output/sam_files/", row$SampleName, ".sam", sep = ""), ">", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".bam", sep = "")), stdout = TRUE, stderr = TRUE)
  system2(command = "samtools", args = c("sort", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".bam", sep = ""), "-o", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".sorted.bam", sep = "")), stdout = TRUE, stderr = TRUE)
  system2(command = "samtools", args = c("index", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".sorted.bam", sep = "")), stdout = TRUE, stderr = TRUE)
  }, error = function(e) {paste("hisat2...", "Alignment failed with error:", e$message)
})
}
```

# Question 5

```{r Alignment_Statistics}
hisat2_logs_dir <- "~/Labs/BIOL4315_Lab4/output/bam_files"
log_files <- list.files(hisat2_logs_dir, pattern = "\\_hisat2\\.log$", full.names = TRUE)
percent_aligned <- 1:length(log_files)
for (i in seq_along(percent_aligned)) {
  percent_aligned[i] <- readLines(log_files[i])[length(readLines(log_files[i]))]
}
align_df <- data.frame(sort(meta_data$SampleName),percent_aligned)
align_df <- align_df %>% 
  mutate(percent_aligned = as.numeric(
    stringr::str_split_i(align_df$percent_aligned, "%",1))) %>% 
  rename(samplename = sort.meta_data.SampleName.)
datatable(align_df)
ggplot()+geom_boxplot(aes(y=align_df$percent_aligned))+scale_x_discrete()+labs(title = "Alignment of A. thaliana Transcript Samples", y = "Percentage Reads Aligned")
```
# Question 6

```{r Feature_Counts}
bfiles <- list.files("~/Labs/BIOL4315_Lab4/output/bam_files", pattern = ".sorted.bam$", full.names = TRUE)
gene_count_list <- Rsubread::featureCounts(files = bfiles, annot.ext = "~/Labs/BIOL4315_Lab4/data/GCF_000001735.4_TAIR10.1_genomic.gtf", isGTFAnnotationFile = TRUE, allowMultiOverlap = FALSE, isPairedEnd = FALSE, nthreads = 8, minMQS = 10, GTF.featureType ="exon", GTF.attrType = "gene_id")
count_table = gene_count_list$counts
new_column_names <- gsub(".sorted.bam", "", colnames(count_table))
colnames(count_table) <- new_column_names
mask = rowSums(count_table) != 0
count_table <- count_table[mask, ]
datatable(count_table)
```
# Question 7

```{r Comparisons}
comp <- systemPipeR::readComp("~/R/x86_64-pc-linux-gnu-library/4.5/systemPipeRdata/extdata/param/targetsPE.txt")
comp[[1]]
```

```{r Differential_Expression_Over_Time}
coldata <- meta_data %>% dplyr::select(SampleName,SampleLong,Factor) %>% 
  dplyr::mutate(SampleLong=str_split_i(SampleLong, "\\.",1)) %>%
  dplyr::rename(condition = SampleLong) %>%
  dplyr::mutate(condition = factor(condition)) %>% 
  dplyr::mutate(Factor = factor(Factor)) 
base::rownames(coldata) <- coldata$SampleName
coldata <- coldata %>% mutate(SampleName = factor(SampleName))
coldata$type <- factor(rep("single-read", nrow(coldata)))
coldata <- coldata[base::match(base::colnames(count_table), rownames(coldata)),]
all(rownames(coldata) == base::colnames(count_table))

dds2 <- DESeqDataSetFromMatrix(countData = count_table, colData = coldata, design = ~ Factor)
dds2_results <- DESeq(dds2)
res2 <- DESeq2::results(dds2_results)

res_M1_A1 <- DESeq2::results(dds2_results, contrast = c("Factor", "M1", "A1"), alpha = 0.2)
res_M1_V1 <- DESeq2::results(dds2_results, contrast = c("Factor", "M1", "V1"), alpha = 0.2)
res_A1_V1 <- DESeq2::results(dds2_results, contrast = c("Factor", "A1", "V1"), alpha = 0.2)
res_M6_A6 <- DESeq2::results(dds2_results, contrast = c("Factor", "M6", "A6"), alpha = 0.2)
res_M6_V6 <- DESeq2::results(dds2_results, contrast = c("Factor", "M6", "V6"), alpha = 0.2)
res_A6_V6 <- DESeq2::results(dds2_results, contrast = c("Factor", "A6", "V6"), alpha = 0.2)
res_M12_A12 <- DESeq2::results(dds2_results, contrast = c("Factor", "M12", "A12"), alpha = 0.2)
res_M12_V12 <- DESeq2::results(dds2_results, contrast = c("Factor", "M12", "V12"), alpha = 0.2)
res_A12_V12 <- DESeq2::results(dds2_results, contrast = c("Factor", "A12", "V12"), alpha = 0.2)

filter_and_count <- function(res_obj, comparison_name, fc_threshold = 2) {
  res_filtered <- res_obj[!is.na(res_obj$padj) & !is.na(res_obj$log2FoldChange), ]
  sig_genes <- res_filtered[abs(res_filtered$log2FoldChange) >= log2(fc_threshold), ]
  up_regulated <- sum(sig_genes$log2FoldChange > 0)
  down_regulated <- sum(sig_genes$log2FoldChange < 0)
  return(data.frame(
    Comparison = comparison_name,
    Up_regulated = up_regulated,
    Down_regulated = down_regulated
  ))
}

results_summary <- rbind(
  filter_and_count(res_M1_A1, "M1 vs A1"),
  filter_and_count(res_M1_V1, "M1 vs V1"),
  filter_and_count(res_A1_V1, "A1 vs V1"),
  filter_and_count(res_M6_A6, "M6 vs A6"),
  filter_and_count(res_M6_V6, "M6 vs V6"),
  filter_and_count(res_A6_V6, "A6 vs V6"),
  filter_and_count(res_M12_A12, "M12 vs A12"),
  filter_and_count(res_M12_V12, "M12 vs V12"),
  filter_and_count(res_A12_V12, "A12 vs V12")
)
print(results_summary)


```

```{r DE_Over_Time_Visual}
plot_data <- results_summary %>%
  pivot_longer(cols = c(Up_regulated, Down_regulated), 
               names_to = "Regulation", 
               values_to = "Count") %>%
  mutate(Regulation = factor(Regulation, levels = c("Up_regulated", "Down_regulated")))

p <- ggplot(plot_data, aes(x = Comparison, y = Count, fill = Regulation)) +
  geom_bar(stat = "identity", position = "stack") +
  coord_flip() +  # Makes it horizontal
  labs(
    title = "Differentially Expressed Genes by Comparison",
    subtitle = "Fold Change >= 2, alpha = 0.2",
    x = "Comparison",
    y = "Number of Genes",
    fill = "Regulation"
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
    plot.subtitle = element_text(hjust = 0.5, size = 12),
    axis.text = element_text(size = 10),
    axis.title = element_text(size = 12),
    legend.title = element_text(size = 11),
    legend.text = element_text(size = 10))
p
```


