---
title: "Lab4_Transcriptomics"
author: "Mendel Kwan"
date: "2025-09-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r Required_Libraries, include = FALSE}

lapply(c("docopt", "DT", "pheatmap","GenomicFeatures", "DESeq2", "edgeR", "systemPipeR", "systemPipeRdata", "BiocStyle", "GO.db", "dplyr", "tidyr", "stringr", "Rqc", "QuasR", "DT", "ape", "clusterProfiler","biomaRt","EnhancedVolcano"), require, character.only = TRUE)
library("DT")
```
# Question 1

```{r Locate_Samples}
fq_path <- systemPipeRdata::pathList()$fastqdir
meta_data <- read.table("/home/vboxuser/R/x86_64-pc-linux-gnu-library/4.5/systemPipeRdata/extdata/param/targetsPE.txt", header = TRUE)
meta_data$FileName1 <- gsub("./data/", fq_path, meta_data$FileName1)
meta_data$FileName2<- gsub("./data/", fq_path, meta_data$FileName2)
datatable(meta_data)
```

# Question 2

```{r Quality_Control_Report}
qcReport <- rqc(fq_path, pattern = ".fastq.gz", openBrowser = FALSE)
rqcCycleQualityBoxPlot(qcReport[1:12])
rqcCycleQualityBoxPlot(qcReport[13:24])
rqcCycleQualityBoxPlot(qcReport[25:36])
rqcCycleBaseCallsLinePlot(qcReport[1:12])
rqcCycleBaseCallsLinePlot(qcReport[13:24])
rqcCycleBaseCallsLinePlot(qcReport[25:36])
```

None of the samples have much longer reads than the others.  They are all between 75-100 bases.  The shortest number of cycles is 75.

None of the samples have a significant number of N base calls.

# Question 3

```{r Quality_Control}
fastqFiles = c()
outputFiles = c()
output_path = "~/Labs/BIOL4315_Lab4/data/processed_reads/"
for(i in 1:nrow(meta_data)) {
  row <- meta_data[i, ]
  fastqFiles <- c(fastqFiles, row$FileName1, row$FileName2)
  outputFileName1 <- gsub("/home/vboxuser/R/x86_64-pc-linux-gnu-library/4.5/systemPipeRdata/extdata/fastq/", output_path, row$FileName1)
  outputFileName1 <- gsub("fastq.gz", "processed.fastq", outputFileName1)
  meta_data[i, "FileName1"] = outputFileName1
  outputFileName2 <- gsub("/home/vboxuser/R/x86_64-pc-linux-gnu-library/4.5/systemPipeRdata/extdata/fastq/", output_path, row$FileName2)
  outputFileName2 <- gsub("fastq.gz", "processed.fastq", outputFileName2)
  meta_data[i, "FileName2"] <- outputFileName2
  outputFiles <- c(outputFiles, outputFileName1, outputFileName2 )
}

preprocessReads(fastqFiles, outputFiles, nBases = 1, truncateEndBases = 3, Lpattern = "GCCCGGGTAA")
datatable(meta_data)

```

```{r Index_Reference, include = FALSE}
reference_genome <- "~/Labs/BIOL4315_Lab4/data/GCF_000001735.4_TAIR10.1_genomic.fna"
tryCatch({
  system2(command = "hisat2-build", args = c("-p","8", reference_genome, "~/Labs/BIOL4315_Lab4/data/hisat2_index/tair10.1_index"), stdout = TRUE, stderr = TRUE)
  }, error = function(e) {
paste("hisat2-build...", "Indexing failed with error:", e$message)
})
```

hisat2-build -p 8 ~/Labs/BIOL4315_Lab4/data/GCF_000001735.4_TAIR10.1_genomic.fna ~/Labs/BIOL4315_Lab4/data/hisat2_index/tair10.1_index

# Question 4
```{r Alignment}
for(i in 1:nrow(meta_data)) {
row <- meta_data[i, ]
tryCatch({
  hisat2_output <- system2(command = "hisat2", args = c("-x","~/Labs/BIOL4315_Lab4/data/hisat2_index/tair10.1_index", row$FileName1, "-p", "8", "-S", paste("~/Labs/BIOL4315_Lab4/output/sam_files/", row$SampleName, ".sam", sep = "" )), stdout = TRUE, stderr = TRUE)
  writeLines(hisat2_output, paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, "_hisat2.log", sep = "") )
  system2(command = "samtools", args = c("view", "-bS", paste("~/Labs/BIOL4315_Lab4/output/sam_files/", row$SampleName, ".sam", sep = ""), ">", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".bam", sep = "")), stdout = TRUE, stderr = TRUE)
  system2(command = "samtools", args = c("sort", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".bam", sep = ""), "-o", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".sorted.bam", sep = "")), stdout = TRUE, stderr = TRUE)
  system2(command = "samtools", args = c("index", paste("~/Labs/BIOL4315_Lab4/output/bam_files/", row$SampleName, ".sorted.bam", sep = "")), stdout = TRUE, stderr = TRUE)
  }, error = function(e) {paste("hisat2...", "Alignment failed with error:", e$message)
})
}
```

# Question 5


